### PR TITLE
Safeguard

### LINKED ISSUE #1084: HTTP/2 does not handle killed requests
##### Steps to reproduce the problem:

1. `mitmdump -S test_replay --kill`
2. Open a website that is not contained in the replay file.

##### What went wrong?

```
127.0.0.1:50907: clientconnect
127.0.0.1:50906: clientconnect
127.0.0.1 GET https://news.ycombinator.com/ HTTP/2.0
 << Connection killed
Exception in thread Thread-2:
Traceback (most recent call last):
  File "c:\python27\Lib\threading.py", line 810, in __bootstrap_inner
    self.run()
  File "c:\users\user\git\mitmproxy\mitmproxy\protocol\http2.py", line 422, in run
    layer()
  File "c:\users\user\git\mitmproxy\mitmproxy\protocol\http.py", line 188, in __call__
    self.process_request_hook(flow)
  File "c:\users\user\git\mitmproxy\mitmproxy\protocol\http.py", line 356, in process_request_hook
    raise Kill()
Kill
```

### PR DESCRIPTION
**Pull request by Kriechi**
prevent backtrace with lost http2 connections
might fix #1084

### CODE REVIEW

File: mitmproxy/protocol/http.py [hunk 1]
```diff
@@ -359,7 +364,10 @@ def establish_server_connection(self, flow):
                 self.set_server(address, tls, address.host)
             # Establish connection is neccessary.
             if not self.server_conn:
-                self.connect()
```
Comment (mhils) [inline mitmproxy/protocol/http.py]: We need to fail if there is an error with the server connection. Just returning False is not an option I think.

Comment (Kriechi) [inline mitmproxy/protocol/http.py]: How hard would you want it to fail? Throwing an exception?

Comment (mhils) [inline mitmproxy/protocol/http.py]: Yes. We should talk about this later or tomorrow, I actually disagree with all changes to `http.py`. :innocent:

**Comment by Kriechi** [comment]
ping @mhils

File: mitmproxy/protocol/http.py [hunk 2]
```diff
@@ -304,6 +304,11 @@ def get_response():
             # > server detects timeout, disconnects
             # > read (100-n)% of large request
             # > send large request upstream
+
+            if "Http2Layer" in [repr(l) for l in self.layers]:
```
Comment (mhils) [inline mitmproxy/protocol/http.py]: Should we rather check for Http2Exception here?

**Comment by mhils** [comment]
Other than my comment above (and the merge conflict), this looks awesome. :+1: :smiley:

**Comment by Kriechi** [comment]
Done.

**Comment by mhils** [comment]
:tada:

**Pull request merged**